### DISCUSSION TITLE
Pico W, uselect.select and uselect.poll blocking intermittently when timeout specified

### DISCUSSION THREAD

**Discussion by gmrza** [micropython/micropython#9480] [Libraries & Drivers]
Hi,

I am running into some issues where I am seeing uselect.select and uselect.poll (or ipoll) blocking even when I specify a timeout in the call.  This behaviour is intermittent, with occurrences sometimes minutes apart, but generally hours apart.
What is helping as a workaround is to set the socket as non-blocking as well, by setting a timeout on the socket.  (As a rule I am setting the timeout on the socket to be 500ms longer than the timeout in my call to poll or select.  I then occasionally get an Exception - ETIMEOUT.)

Has anybody else seen this behaviour?

Regards,
Geoff.

**Comment by dpgeorge** [comment]
> I am running into some issues where I am seeing uselect.select and uselect.poll (or ipoll) blocking even when I specify a timeout in the call.

How long does the call to select/poll block for, beyond the specified timeout?  Does it block forever?

Using `socket.socket` objects in select/poll should never block.  It's simply a loop that keeps polling until the socket is ready.

Can you share code that shows the problem?

**Reply by gmrza** [reply]
From what I can see, once it blocks, it blocks forever.  What I should probably do before sharing the code is create a version that cuts out a lot of other sections, to try to focus just on this issue.  I would rather do that first.
One other weird thing I have noticed is that it only seems to happen outdoors, where my WiFi connectivity is a weaker.
Seen from the AP, the Pico W outdoors has a signal strength of -71dBm compared to -65dBm for the one that is indoors.

**Reply by dpgeorge** [reply]
It should not block forever.  There must be something wrong, eg maybe the whole board locks up (are you certain it blocks in the select/poll call?).

**Reply by gmrza** [reply]
I'm pretty certain it is locking in the select() statement.  There is one other point where I am checking for a timeout on an accept() statement, and I am generating the same error message.  I can update my logging so that I can be certain which line it is blocking in.

**Reply by gmrza** [reply]
I've got an updated version of the code running on my desktop dev board that has try: statements only protecting the select(), accept() and recv() statements, so that anything else that generates an exception will cause it to reset the board, and my error logging will tell me which statement caused the exception.  I am pretty sure it is the select() statement, but I will find out soon.

**Reply by gmrza** [reply]
I've tracked it now to the recv() statement, which shouldn't block, given the the select() statement has returned the fact that that socket is ready for reading.  What I have noticed as well is that this is happening the most frequently when the client with the web browser is also outside.  This may thus be related to packet loss.  Let me run a tcpdump to see what is happening.

**Reply by dpgeorge** [reply]
If the socket is non-blocking, `socket.recv()` will never block.  If the socket is ready for reading, it should also never block.  So something else is going on here.

We'd need to see the code to get more insight.

**Reply by gmrza** [reply]
I've pulled out the lines of code relating to the socket use.

Initial setup, before we go into the loop is as follows:
```
addr = socket.getaddrinfo('0.0.0.0', 80)[0][-1]

s = socket.socket()
s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
s.settimeout(4.5)
s.bind(addr)
s.listen(1)
```

the main loop runs infinitely (while True).
The socket code is as follows:

First:
```
        try:
            r, w, x = uselect.select([s], [],[], timeout) # timeout is a value from 1 to 4 seconds
        except OSError as e:
            l.append('Select: {}'.format(e))
            print(e)
            r = [] # Make sure r is defined
```
Then there is a block which reads a BME280 sensor via I2C

And then we go on to read data from the socket if len(r) > 0:
```
        if len(r) > 0:
            try:
                cl, addr = r[0].accept()
            except OSError as e:
                l.append('Accept: {}'.format(e))
                print('Accept: {}'.format(e))
            else:
                #print('client connected from', addr)
                try:
                    request = cl.recv(1024)
                except OSError as e:
                    l.append('Recv: {}'.format(e))
                    print('Recv: {}'.format(e))
                else:
                    #If we didn't raise an exception, process the data read
```

The only line that times out is the recv() call, which shouldn't time out, as it is only called if there is data to be read.  As I've noted earlier I am only really seeing this behaviour if my WiFi connectivity is not so great.  What is also appearing to be the case is that it is more likely to occur when using a Windows or Android client than when using curl or a browser on Mac OS - I don't have data to confirm that however.

**Reply by dpgeorge** [reply]
> The only line that times out is the recv() call, which shouldn't time out, as it is only called if there is data to be read.

I don't think that's true, there's not necessarily data to read.  The select will return readable because the socket `s` has a waiting, incoming connection.  That means the call to `accept()` will succeed without blocking.  But then you need to do another call to `select()` on the *client socket `cl`* before you can call `cl.recv(1024)` without blocking.

**Reply by gmrza** [reply]
Thanks.  That seems to be making a difference, at least in a good connectivity environment.  I'll need to deploy the update outdoors to really test it.

**Reply by gmrza** [reply]
I think mobile browsers seem to do things a bit differently, which is was also causing part of my problem.  I've now modified my code to use a state machine to take it through the steps on accepting and reading from the connection, while keeping my loop to read the sensors running.  Once I tidy up my code that will provide a more elegant solution.
Of course once I move to using asyncio and mqtt this will all be moot....

**Reply by dpgeorge** [reply]
Moving to `asyncio` would probably make a lot of these problems disappear, because the built-in socket server in `asyncio` will handle all this for you.

**Comment by peterhinch** [comment]
The fact that your problem occurs when RSSI is low suggests that you may be encountering brief WiFi outages.

Writing applications capable of handling WiFi outages is quite difficult - see [these notes](https://github.com/peterhinch/micropython-samples/blob/master/resilient/README.md). The approach we used in developing [asynchronous MQTT](https://github.com/peterhinch/micropython-mqtt/blob/master/mqtt_as/README.md) differed from yours in that we used nonblocking sockets rather than `select.poll` but some of the observations may be relevant.

**Reply by gmrza** [reply]
I've got my code to a point where it is working reliably enough.  I think you may have helped with the next step.  My intention was to move to MQTT - I needed to work through the learning curve of learning to work with MicroPython and the Pico W first.  I also have an existing InfluxDB/Grafana setup that I needed to fit into.  Using MQTT would be a more elegant way to get data into InfluxDB, however.  The links you shared will probably be a big help to getting me on my way.
